public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -0,0 +1,102 @@
+import { Action } from '@reduxjs/toolkit';
+import { useCallback } from 'react';
+
+import { dispatch, getState } from 'app/store/store';
+import { RuleGroupIdentifier } from 'app/types/unified-alerting';
+import { PostableRuleDTO, RulerGrafanaRuleDTO, RulerRuleDTO } from 'app/types/unified-alerting-dto';
+
+import { alertRuleApi } from '../api/alertRuleApi';
+import { addRuleAction, deleteRuleAction, pauseRuleAction, ruleGroupReducer } from '../reducers/ruler/ruleGroups';
+import {
+  fetchPromAndRulerRulesAction,
+  fetchRulesSourceBuildInfoAction,
+  getDataSourceRulerConfig,
+} from '../state/actions';
+
+export function useProduceNewRuleGroup() {
+  const [fetchRuleGroup, _fetchRuleGroupState] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
+  const [deleteRuleGroup, _deleteRuleGroupState] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useLazyQuery();
+  const [updateRuleGroup, updateRuleGroupState] = alertRuleApi.endpoints.updateRuleGroupForNamespace.useMutation();
+
+  const produceNewRuleGroup = async (ruleGroup: RuleGroupIdentifier, action: Action) => {
+    const { ruleSourceName, group, namespace } = ruleGroup;
+
+    // @TODO we should really not work with the redux state (getState) here
+    await dispatch(fetchRulesSourceBuildInfoAction({ rulesSourceName: ruleSourceName }));
+    const rulerConfig = getDataSourceRulerConfig(getState, ruleSourceName);
+
+    const currentRuleGroup = await fetchRuleGroup({ rulerConfig, namespace, group }).unwrap();
+
+    // @TODO convert rule group to postable rule group – TypeScript is not complaining here because
+    // the interfaces are compatible but it _should_ complain
+    const newRuleGroup = ruleGroupReducer(currentRuleGroup, action);
+
+    // if this was the last rule in the group after applying the action on the reducer we remove the group in its entirety
+    const deleteEntireGroup = newRuleGroup.rules.length === 0;
+    if (deleteEntireGroup) {
+      return deleteRuleGroup({
+        rulerConfig,
+        namespace,
+        group,
+      });
+    }
```

Comment:
This feels out of place here, feels like it should sit within the `delete` logic - so you (conceptually) are only going to worry about deleting everything when you're deleting a rule. Or is there another way that the group can be emptied out without deleting?

Reply:
Resolved offline, but we struggled to find a good pattern that allowed the logic to live in `useDeleteRuleFromGroup` whilst also preventing two separate HTTP requests from being made.